File: src/lib/columnRegistry.ts
// Column Registry - collects all unique columns from all towers
// Each tower may have different columns, missing columns get NaN

export class ColumnRegistry {
  private columns: Set<string> = new Set(['SECONDS', 'NANOSECONDS']);
  private towerColumns: Map<string, Set<string>> = new Map(); // site -> columns
  
  // Add columns from a single tower
  addTowerColumns(siteName: string, columns: string[]): void {
    const siteCols = new Set(columns);
    this.towerColumns.set(siteName, siteCols);
    
    // Merge into global registry
    for (const col of columns) {
      this.columns.add(col);
    }
  }
  
  // Get all unique columns across all towers
  getAllColumns(): string[] {
    return Array.from(this.columns).sort();
  }
  
  // Get columns for a specific tower
  getTowerColumns(siteName: string): string[] {
    return Array.from(this.towerColumns.get(siteName) || []);
  }
  
  // Check if a column exists for a specific tower
  hasColumn(siteName: string, columnName: string): boolean {
    return this.towerColumns.get(siteName)?.has(columnName) || false;
  }
  
  // Get the registry as a JSON-serializable object
  toJSON(): { allColumns: string[]; towerColumns: Record<string, string[]> } {
    const towerCols: Record<string, string[]> = {};
    for (const [site, cols] of this.towerColumns) {
      towerCols[site] = Array.from(cols);
    }
    
    return {
      allColumns: this.getAllColumns(),
      towerColumns: towerCols,
    };
  }
  
  // Load from JSON
  static fromJSON(data: { allColumns: string[]; towerColumns: Record<string, string[]> }): ColumnRegistry {
    const registry = new ColumnRegistry();
    for (const col of data.allColumns) {
      registry.columns.add(col);
    }
    for (const [site, cols] of Object.entries(data.towerColumns)) {
      registry.towerColumns.set(site, new Set(cols));
    }
    return registry;
  }
}

// Singleton instance
export const columnRegistry = new ColumnRegistry();

// Get comprehensive list of all possible DAQM columns
export function getAllDaqmColumns(): string[] {
  return columnRegistry.getAllColumns();
}

// Check if a site has a specific column
export function siteHasColumn(siteName: string, columnName: string): boolean {
  return columnRegistry.hasColumn(siteName, columnName);
}

File: React_FastAPI_Live_EC/backend/column_registry.py
"""Column Registry — tracks all unique columns from all towers."""


class ColumnRegistry:
    def __init__(self):
        self._columns: set[str] = {"SECONDS", "NANOSECONDS"}
        self._tower_columns: dict[str, set[str]] = {}

    def add_tower_columns(self, site_name: str, columns: list[str]):
        site_cols = set(columns)
        self._tower_columns[site_name] = site_cols
        for col in columns:
            self._columns.add(col)

    def get_all_columns(self) -> list[str]:
        return sorted(self._columns)

    def get_tower_columns(self, site_name: str) -> list[str]:
        return sorted(self._tower_columns.get(site_name, []))

    def has_column(self, site_name: str, column_name: str) -> bool:
        return column_name in self._tower_columns.get(site_name, set())

    def to_dict(self) -> dict:
        return {
            "allColumns": self.get_all_columns(),
            "towerColumns": {site: sorted(cols) for site, cols in self._tower_columns.items()},
        }

    @staticmethod
    def from_dict(data: dict) -> "ColumnRegistry":
        registry = ColumnRegistry()
        for col in data.get("allColumns", []):
            registry._columns.add(col)
        for site, cols in data.get("towerColumns", {}).items():
            registry._tower_columns[site] = set(cols)
        return registry


# Singleton instance
column_registry = ColumnRegistry()
